### PR TITLE
Datapar zip iterator

### PR DESCRIPTION
**Pull request by hkaiser**
This provides a massive simplification of implementing a vectorization backend for the parallel algorithms. Essentially we now require the implementation of four relatively small traits for each of the back ends.

This patch also adds support for zip_iterators. This will simplify to adapt many of the remaining algorithms for vectorization.

### CODE REVIEW

**Comment by hkaiser** [comment]
Please note, that while the files to support LibFlatArray as a backend are in place, the current code does not compile yet because of missing interfaces in the back end library itself. The othe back end implementations (Vc and Boost.Simd) work as expected.

File: hpx/parallel/datapar/zip_iterator.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        call(hpx::util::zip_iterator<Iter...> const& it,
+            hpx::util::detail::pack_c<std::size_t, Is...>)
+        {
+            auto const& t = it.get_iterator_tuple();
+            int const sequencer[] = {
+                0, (data_alignment(hpx::util::get<Is>(t)), 0)...
+            };
+            return std::accumulate(
+                &sequencer[0], &sequencer[sizeof(sequencer)/sizeof(sequencer[0])],
+                std::size_t(0));
```
Comment (sithhell) [inline hpx/parallel/datapar/zip_iterator.hpp]: Is the alignment really the accumulation of all? Isn't it rather the maximum?

Comment (hkaiser) [inline hpx/parallel/datapar/zip_iterator.hpp]: This is assuming that the data referenced by the overall `zip_iterator` is 'properly' aligned only if the data referenced by all embedded iterators are 'properly' aligned. In this context, the function will return zero (i.e. is 'properly' aligned) only iof all separate data ranges are aligned.

I should probably rename this trait to `is_aligned` or somesuch instead.

**Comment by hkaiser** [comment]
@sithhell The latest change addresses your comment.

File: hpx/parallel/datapar/zip_iterator.hpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    struct is_data_aligned_impl<hpx::util::zip_iterator<Iter...> >
+    {
+        template <std::size_t ... Is>
+        static HPX_FORCEINLINE bool
+        call(hpx::util::zip_iterator<Iter...> const& it,
+            hpx::util::detail::pack_c<std::size_t, Is...>)
+        {
+            auto const& t = it.get_iterator_tuple();
+            int const sequencer[] = {
+                0, (is_data_aligned(hpx::util::get<Is>(t)), 0)...
```
Comment (hkaiser) [inline hpx/parallel/datapar/zip_iterator.hpp]: Yah, we could use `std::any_of()` over a sequence of `bool`s. That might even be (slightly) faster.

**Comment by sithhell** [comment]
The comment was not meant to be about performance, more about code clarity
;)

**Pull request merged**